probe-rs/src/flashing/builder.rs

```diff
@@ -0,0 +1,697 @@
+                        if let Some(page_info) = page_info {
+                            let new_page = FlashPage::new(&page_info);
+                            pages.push(new_page.clone());
+                            log::trace!(
+                                "Added Page (0x{:08x}..0x{:08x})",
+                                page_info.base_address,
+                                page_info.base_address + page_info.size
+                            );
+                        } else {
+                            return Err(FlashError::InvalidFlashAddress(flash_address));
+                        }
+                        continue;
+                    }
+                } else {
+                    // If no sector exists, create a new one.
+                    log::trace!("Trying to create a new sector");
+                    let sector_info = flash_algorithm.sector_info(flash_address);
+
+                    if let Some(sector_info) = sector_info {
+                        let new_sector = FlashSector::new(&sector_info);
+                        sectors.push(new_sector);
+                        log::debug!(
+                            "Added Sector (0x{:08x}..0x{:08x})",
+                            sector_info.base_address,
+                            sector_info.base_address + sector_info.size
+                        );
+                    } else {
+                        return Err(FlashError::InvalidFlashAddress(flash_address));
+                    }
+                    continue;
+                }
+            }
+        }
+
+        // Fill the page gap if there is one.
+        if let Some(page) = pages.last_mut() {
+            fill_page(page)?;
+        }
+
+        log::debug!("Sectors are:");
+        for sector in &sectors {
+            log::debug!("{:#?}", sector);
+        }
+
+        log::debug!("Pages are:");
+        for page in &pages {
+            log::debug!("{:#?}", page);
+        }
+
+        Ok(FlashLayout {
+            sectors,
+            pages,
+            fills: vec![],
+        })
+    }
+
+    pub(super) fn add_sector<'b>(
+        &self,
+        flash_algorithm: &FlashAlgorithm,
+        address: u32,
+        sectors: &'b mut Vec<FlashSector>,
+    ) -> Result<&'b mut FlashSector, FlashError> {
+        let sector_info = flash_algorithm.sector_info(address);
+        if let Some(sector_info) = sector_info {
+            let new_sector = FlashSector::new(&sector_info);
+            sectors.push(new_sector);
+            log::trace!(
+                "Added Sector (0x{:08x}..0x{:08x})",
+                sector_info.base_address,
+                sector_info.base_address + sector_info.size
+            );
+            // We just added a sector, so this unwrap can never fail!
+            Ok(sectors.last_mut().unwrap())
+        } else {
+            Err(FlashError::InvalidFlashAddress(address))
+        }
+    }
+
+    pub(super) fn add_page<'b>(
+        &self,
```

Comment:
Here as well, `self` is never used. Could we put the pages into the `FlashBuilder`?